### ISSUE TITLE
Flaky test `manipulation_tests.test_basic.TestBasic.test_copyto_multigpu`

### ISSUE THREAD

**Issue by emcastillo** [cupy/cupy#4673]
```
Error Message
AssertionError: 
Arrays are not equal

Mismatched elements: 24 / 24 (100%)
Max absolute difference: 23.
Max relative difference: 1.
 x: array([[[0.   , 0.   , 0.   , 1.984],
        [0.   , 0.   , 0.   , 2.   ],
        [0.   , 0.   , 0.   , 2.016]],...
 y: array([[[ 1.,  2.,  3.,  4.],
        [ 5.,  6.,  7.,  8.],
        [ 9., 10., 11., 12.]],...
```

https://jenkins.preferred.jp/job/chainer/job/cupy_pr/3016/TEST=cupy-py3-cub,label=mn1-p100/testReport/junit/tests.cupy_tests.manipulation_tests.test_basic/TestBasic/test_copyto_multigpu/

**Comment by kmaehashi** [comment]
Other cases:
https://jenkins.preferred.jp/job/chainer/job/cupy_pr/TEST=cupy-py3-cub,label=mn1-p100/3005/testReport/
https://jenkins.preferred.jp/job/chainer/job/daily_master_cupy-shuffled/449/ID=14,label=mn1-p100/testReport/junit/tests.cupy_tests.manipulation_tests.test_basic/TestBasic/test_copyto_multigpu/

**Comment by kmaehashi** [comment]
This test failures shows a similar behavior, although this is not a multi-gpu test.
https://jenkins.preferred.jp/job/chainer/job/cupy_pr/3044/TEST=cupy-py3,label=mn1-p100/testReport/junit/tests.cupy_tests.cuda_tests.test_texture/TestCUDAarray_param_25_%7Bdimensions=(67,%200,%200),%20dtype=int8,%20n_channels=1,%20stream=True,%20xp='cupy'%7D/test_array_gen_cpy/

**Comment by leofang** [comment]
Could it be something wrong with the random numbers?

**Comment by kmaehashi** [comment]
As for `test_copyto_multigpu `, I think random numbers are not related, as it just tries to copy `arange` result between devices.

**Comment by takagi** [comment]
I cannot reproduce this behavior on my local 😢

**Comment by leofang** [comment]
Just a thought: fixing #3980 might fix this, as the "current stream" shouldn't be the same for all devices.

**Comment by leofang** [comment]
> btw, does the CI have nvlinks connecting the gpus? I can't reproduce this either...

Needs to do a stress test similar to what's done in https://github.com/cupy/cupy/issues/4865#issuecomment-804272731: Open two terminals locally, and run the following script simultaneously in each terminal:
```python
import numpy as np
import cupy as cp
from cupy import cuda
from cupy import testing


def test(xp, dtype):
    with cuda.Device(0):
        s = cuda.Stream()
        s.use()
        a = testing.shaped_arange((2, 3, 4), xp, dtype)
    with cuda.Device(1):
        b = xp.empty((2, 3, 4), dtype=dtype)
    xp.copyto(b, a)
    return b


for i in range(10000):
    s = cp.cuda.Stream()
    with s:
        for dtype in '?bhilqBHILQefdFD':
        #for dtype in 'f':
            out = [None]*2
            for i, xp in enumerate((np, cp)):
                out[i] = test(xp, dtype)
            testing.assert_allclose(out[0], out[1])                                                 
```

**Comment by leofang** [comment]
> Just a thought: fixing #3980 might fix this, as the "current stream" shouldn't be the same for all devices.

Unlike #4946 which is relatively straightforward to fix, I think from the above reproducer that the work proposed in #3980 is definitely needed to make a proper fix, because it is not at all obvious in the failing test code
https://github.com/cupy/cupy/blob/d8e18f8c8ebb2e5f5f6147da5ea2fd2047274d24/tests/cupy_tests/manipulation_tests/test_basic.py#L82-L86
that one should expect a per-device stream issue. 

Note that In my reproducer I added extra streams and stressed the hardware to make the bug appear; missing any ingredient would not reproduce it.

**Comment by leofang** [comment]
Even better reproducer to stress the need of a per-device current stream:
```python
import cupy as cp
import pytest


for i in range(100):
    s = cp.cuda.Stream()
    s.use()
    status = pytest.main(['tests/cupy_tests/manipulation_tests/test_basic.py', '-k', 'test_copyto_multigpu'])
    if status != 0:
        raise RuntimeError(i, status)
```
which leads to
```
E   cupy_backends.cuda.api.driver.CUDADriverError: CUDA_ERROR_INVALID_HANDLE: invalid resource handle
```

**Comment by leofang** [comment]
I enabled per-device current streams locally (https://github.com/leofang/cupy/tree/per_device_stream), and `CUDA_ERROR_INVALID_HANDLE` from the above script is gone, but the actual multi-GPU copy tests still fail under the stress test. Not sure what I am missing...

**Comment by emcastillo** [comment]
> I enabled per-device current streams locally (https://github.com/leofang/cupy/tree/per_device_stream), and CUDA_ERROR_INVALID_HANDLE from the above script is gone, but the actual multi-GPU copy tests still fail under the stress test. Not sure what I am missing...

Maybe we need to synchronize both streams?

**Comment by leofang** [comment]
I thought about that too, but I tried this patch to establish a correct stream order and it still fails...
```diff
diff --git a/cupy/_manipulation/basic.py b/cupy/_manipulation/basic.py
index c75c7d501..90c4b3a9a 100644
--- a/cupy/_manipulation/basic.py
+++ b/cupy/_manipulation/basic.py
@@ -1,6 +1,7 @@
 import numpy
 
 from cupy import core
+from cupy import cuda
 from cupy.core import _fusion_interface
 from cupy.core import fusion
 from cupy._sorting import search
@@ -55,7 +56,13 @@ def copyto(dst, src, casting='same_kind', where=None):
 
     if where is None:
         if _can_memcpy(dst, src):
-            dst.data.copy_from_async(src.data, src.nbytes)
+            with src.device:
+                e = cuda.Event(disable_timing=True)
+                e.record()
+                src_stream = cuda.get_current_stream()
+                src_stream.wait_event(e)
+                print('go!')
+                dst.data.copy_from_async(src.data, src.nbytes, src_stream)
         else:
             device = dst.device
             with device:
```

**Comment by emcastillo** [comment]
can we do the data copy the other way around?

**Comment by leofang** [comment]
We can, and I can give it a shot tomorrow, but I am not sure why it'd help. With my patches, we correctly call `cudaMemcpyAsync` on the source's current stream, so shouldn't everything be serialized on that?

**Comment by kmaehashi** [comment]
> btw, does the CI have nvlinks connecting the gpus? I can't reproduce this either...

CI does not have NVLink.

> https://github.com/cupy/cupy/issues/4673#issuecomment-804335269

Thanks for the reproducer! However, I think this reproducer is not an appropriate for `test_copyto_multigpu` flaky issue. This code enqueues DtoH copy of `b` *after* destroying the stream used for `arange`, so D2H can happen before  `arange` kernel or P2P copy.
![Screen Shot 2021-04-06 at 2 35 30](https://user-images.githubusercontent.com/939877/113605186-422b1780-9681-11eb-8a21-332d1b5b0323.png)

**Comment by leofang** [comment]
Sorry, which reproducer did you refer to? 1 (https://github.com/cupy/cupy/issues/4673#issuecomment-804335269) or 2 (https://github.com/cupy/cupy/issues/4673#issuecomment-804350948)?

**Comment by kmaehashi** [comment]
The first one.

**Comment by kmaehashi** [comment]
For record: this is a regression in v8.0.0a1. https://github.com/cupy/cupy/pull/2942

**Comment by kmaehashi** [comment]
I think the second code is not a reproducer of this issue as well; launching a kernel using a stream that belongs to a different device will raise that error (although I understand that we need to manage stream per device 😃 )

```py
>>> import cupy
>>> cupy.cuda.Stream().use()
<Stream 93952003582992>
>>> cupy.cuda.Device(1).use()
>>> cupy.arange(10)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/maehashi/Development/cupy/cupy/_creation/ranges.py", line 57, in arange
    _arange_ufunc(typ(start), typ(step), ret, dtype=dtype)
  File "cupy/_core/_kernel.pyx", line 1105, in cupy._core._kernel.ufunc.__call__
  File "cupy/cuda/function.pyx", line 215, in cupy.cuda.function.Function.linear_launch
  File "cupy/cuda/function.pyx", line 183, in cupy.cuda.function._launch
  File "cupy_backends/cuda/api/driver.pyx", line 306, in cupy_backends.cuda.api.driver.launchKernel
  File "cupy_backends/cuda/api/driver.pyx", line 125, in cupy_backends.cuda.api.driver.check_status
cupy_backends.cuda.api.driver.CUDADriverError: CUDA_ERROR_INVALID_HANDLE: invalid resource handle
```

**Comment by leofang** [comment]
> However, I think this reproducer is not an appropriate for `test_copyto_multigpu` flaky issue. This code enqueues DtoH copy of `b` _after_ destroying the stream used for `arange`, so D2H can happen before `arange` kernel or P2P copy.

hmmm, the 1st reproducer is taken from the failing test:
https://github.com/cupy/cupy/blob/10a3d9fda11ac90c283976a1bcd1d0e589fd922e/tests/cupy_tests/manipulation_tests/test_basic.py#L78-L87
Are you saying the extra line `s.use()` is not functioning correctly? But isn't the stream `s` put in the thread-local storage after `s.use()`? Could it be the destroyed stream is from the previous run? Maybe I am not following...

**Comment by kmaehashi** [comment]
> Are you saying the extra line `s.use()` is not functioning correctly? But isn't the stream `s` put in the thread-local storage after `s.use()`? Could it be the destroyed stream is from the previous run? Maybe I am not following...

`s` will be destroyed after returning from `test`, because `s.use()` will only keep weakref of `s` to thread-local storage.

**Comment by leofang** [comment]
Right I forgot it's a weakref.

**Comment by leofang** [comment]
OK I think you're right @kmaehashi. So we still need a reproducer...😞

**Comment by leofang** [comment]
I still feel the problem is in the lack of per-device current streams. When coping from device A to B, the copy stream must establish a stream order such that 
1. the data is ready on A's current stream
2. the copy is performed on the copy stream
3. B's current stream waits for the copy stream (by stream wait event, as stream synchronization is not possible (copy stream is from source))

**Comment by emcastillo** [comment]
btw,

I just found that increasing the arrange size, we don't need to run the script in two terminals at the same time

```
def test(xp, dtype):
    with cuda.Device(0):
        s = cuda.Stream()
        s.use()
        a = testing.shaped_arange((20, 30, 40), xp, dtype)
    with cuda.Device(1):
        b = xp.empty((20, 30, 40), dtype=dtype)
    xp.copyto(b, a)
    return b


for i in range(10000):
    print(i)
    for dtype in "?bhilqBHILQefdFD":
        out = [None] * 2
        for i, xp in enumerate((np, cp)):
            out[i] = test(xp, dtype)
        testing.assert_allclose(out[0], out[1])
```

```
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 18136 / 24000 (75.6%)
Max absolute difference: 28555.80025144
Max relative difference: 0.
 x: array([[[1.0000e+00+1.0000e+00j, 2.0000e+00+2.0000e+00j,
         3.0000e+00+3.0000e+00j, ..., 3.8000e+01+3.8000e+01j,
         3.9000e+01+3.9000e+01j, 4.0000e+01+4.0000e+01j],...
 y: array([[[1.0000e+00+1.0000e+00j, 2.0000e+00+2.0000e+00j,
         3.0000e+00+3.0000e+00j, ..., 3.8000e+01+3.8000e+01j,
         3.9000e+01+3.9000e+01j, 4.0000e+01+4.0000e+01j],...
```

This means that device 1 is reading from device 0 before it finished creating the array

**Comment by emcastillo** [comment]
We were playing with this offline and found the following reproducer
```
s = cuda.Stream()
def test(xp, dtype, i):
    shape = (2 * 3 * 4,)
    with cuda.Device(0):
        a = testing.shaped_arange(shape, xp, dtype)
    with cuda.Device(1):
        b = xp.empty(shape, dtype=dtype)
    xp.copyto(b, a)
    return b
for i in range(10000):
    print(i)
    for dtype in "hlqfdFD":
        out = [None] * 2
        with s:
            for j, xp in enumerate((np, cp)):
                out[j] = test(xp, dtype, i)
        testing.assert_allclose(out[0], out[1])
```

We suspected that some test is calling a `stream.use()` and not cleaning it after. And bingo
https://github.com/cupy/cupy/blob/master/tests/cupy_tests/cuda_tests/test_stream.py#L133-L137

If we are correct, once this test fails with #5040 we are going to see the tpds stream being activated by default.

**Comment by kmaehashi** [comment]
Issue resolved via #5004 (and #5042).
